the-collab-lab/tcl-33-smart-shopping-list
# The estimated number of days until the next purchase date is calculated

## Description
When a purchase is recorded, this code estimates the number of days until the next purchase date, calculates it and records it in the database.
The script calculateEstimate from @the-collab-lab/shopping-list-utils is used to calculate the next estimated purchase interval. 

It adds three inputs which are timesPurchased, lastPurchased and nextDateOfPurchase in the firebase and updates them for each grocery item that is checked. If it is an existing grocery item in the database, all **_**three data inputs**_** change. 
If it's a new grocery item, only **_two data inputs_** are changed in the firebase: timesPurchased equals 1 and lastPurchased is set to new Date(). 

Even though we had the code that was updating firebase in App.js, I moved it to estimate.js file since it made more sense and makes it legible.

- I learnt that calculateEstimate only takes three parameters and for daysSinceLastTransaction, it shouldn't be _this.sinceLastDaysTransaction_ nor _data.sinceLastDaysTransaction_ and you should just use it as it is when calling calculateEstimate.
- I also learnt that I had to convert the variable _lastPurchase_  to millisecond(toMillis) which is a method of TimeUnit Class that is used to get the time represented by the TimeUnit object, as the number of Seconds, since midnight UTC on the **1st January 1970.**


## Related Issue
Closes #10 


## Acceptance Criteria

When a user checks a grocery item: 
> - the estimated number of days until the next purchase date is calculated & recorded in the database
> - The number of times a grocery item is purchased is also recorded in the database
> - The last purchase date of the grocery item is recorded in the database

## Type of Changes




|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|  ✓ | :sparkles: New feature     |
|  ✓ | :hammer: Refactoring       |
|    | :100: Add tests            |
|    | :link: Update dependencies |
|    | :scroll: Docs              |

## Updates

### Before


![ezgif com-gif-maker (1)](https://user-images.githubusercontent.com/17056354/142302521-df70c94e-492d-4934-98aa-a21b02d569be.gif)


### After

![ezgif com-gif-maker (2)](https://user-images.githubusercontent.com/17056354/142315017-c6f30c47-ee26-4a23-8122-f58dc8c18606.gif)


## Testing Steps / QA Criteria

- Be on the main branch of tcl-33-smart-shopping-list
- Git checkout af-is-calculate-estimate to switch to the specific branch
- **npm start** to start the application which will open in your browser
- if you have a token, enter and click View List, check few grocery items and you will see the timesPurchased incrementing by 1
- if you don't have a token, create your own shipping list, add some items then click the ViewList button and check the box which will make the item increment by 1
- To get access to **last purchase date** and **next purchase date** of grocery items, log into the firebase console and you will automatically see those data inputs changing anytime you purchase/check a grocery item.
